In [1]:
# LambdaGrader Before File Code
# REMOVE_IN_HTML_OUTPUT
import traceback
import json
from datetime import datetime

grading_start_time = datetime.now()

_graded_result = {
    'grading_finished_at': None,
    'grading_duration_in_seconds': 0,
    'learner_score': 0,
    'total_available': 0,
    'num_test_cases': 0,
    'num_passed_cases': 0,
    'num_failed_cases': 0,
    'results': []
}

is_lambdagrader_env = True

def _record_test_case(test_case_name, did_pass, available_points, message=''):
    global _graded_result
    warning_message = ''
    
    if test_case_name in map(lambda x: x['test_case_name'], _graded_result['results']):
        warning_message = f'[Warning] LambdaGrader: An identical test case name "{test_case_name}" already exists. Test cases with identical test case names will be graded \n\n'

    _graded_result['results'].append({
        'test_case_name': test_case_name,
        'available_points': available_points,
        'points': available_points if did_pass else 0,
        'pass': did_pass,
        'message': warning_message + message,
    })

# City of Chicago Uber/Lyft Vehicles Registered

- 🏆 80 points available
- 🤠 Author: Park (ypark32@illinois.edu)
- ✏️ Last updated on 09/19/2022

---

▶️ First, run the code cell below to import `unittest`, a module used for **🧭 Check Your Work** sections and the autograder.

In [2]:
# DO NOT MODIFY THE CODE IN THIS CELL
import unittest
tc = unittest.TestCase()

---

## 💎 Case overview

There have been four Transportation Network Providers (often called rideshare companies 🚗) licensed to operate in Chicago. These rideshare companies are required to routinely report vehicles, drivers, and trips information to the City of Chicago, which are published to the [Chicago Data Portal](https://data.cityofchicago.org/). The latest vehicles dataset can be downloaded at [registered vehicles dataset](https://data.cityofchicago.org/Transportation/Transportation-Network-Providers-Vehicles/bc6b-sq4u). The original dataset has been preprocessed to fit this case study.

The reporting is done on a monthly basis, as indicated in the `REPORTED_YEAR` and `REPORTED_MONTH` columns. For each registered vehicle at a given month, the following information are provided:

| Column Name       | Description                                                                                                                                                                                                                      |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `REPORTD_YEAR`    | The year in which the vehicle was reported                                                                                                                                                                                       |
| `REPORTED_MONTH`  | The month in which the vehicle was reported                                                                                                                                                                                      |
| `STATE`           | The state of the license plate                                                                                                                                                                                                   |
| `MAKE`            | The make of the vehicle                                                                                                                                                                                                          |
| `MODEL`           | The model of the vehicle                                                                                                                                                                                                         |
| `COLOR`           | The color of the vehicle                                                                                                                                                                                                         |
| `MODEL_YEAR`      | The model year of the vehicle                                                                                                                                                                                                    |
| `NUMBER_OF_TRIPS` | Number of trips provided in this month. Due to the complexities of matching, errors are possible in both directions. Values over 999 are converted to null as suspected error values that interfere with easy data visualization |
| `MULTIPLE_TNPS`   | Whether the vehicle was reported by multiple TNPs in this month. Matching is imperfect so some vehicle records that should have been combined may be separate.                                                                   |

### ⚔️ Your goal

In this case study, you'll use Pandas to explore and analyze over 1 million monthly rideshare vehicle registrations. Below are some of the questions you'll be answering.

- How many total trips did drivers make between 2015-2020?
- How many trips did drivers make per month on average between 2015-2020?
- Do Lincoln drivers make more trips compared to the overall average?
- Do any drivers drive a red Tesla?
- Do any drivers drive a Porsche or a Hummer?
- Are there any vehicles that are brand new?
- Do drivers who drive for two or more rideshare companies (e.g., Uber **and** Lyft) make more trips on average compared to other drivers who only drive for only one rideshare company?
- Which automakers are popular amongst rideshare drivers?
- Which automakers have the highest number of trips on average?
- How many number of trips do drivers of Japanese big three automakers make on average?
- Which make/model are popular amongst rideshare drivers?
- How about colors?

### Notes

- Vehicle registrations that made less than 100 monthly trips are excluded. If you'd like to use the full dataset, you can download the full data (~350 MB) at [registered vehicles dataset](https://data.cityofchicago.org/Transportation/Transportation-Network-Providers-Vehicles/bc6b-sq4u).  
- A vehicle will likely be registered multiple times. As an example, if a driver drives a Camry for 12 months, there will be 12 rows of duplicate vehicle information.

---

### 🎯 Part 1: Import `numpy` and `pandas`

#### 👇 Tasks

- ✔️ Import the following Python packages.
    1. `pandas`: Use alias `pd`.
    2. `numpy`: Use alias `np`.

In [3]:
### BEGIN SOLUTION
import pandas as pd
import numpy as np
### END SOLUTION

#### 🧭 Check your work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [4]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-01"
    _points = 2

    tc.assertTrue('pd' in globals(), 'Check whether you have correctly import Pandas with an alias.')
    tc.assertTrue('np' in globals(), 'Check whether you have correctly import NumPy with an alias.')
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

## 🔭 Explore Data

### 📌 Load data

▶️ Run the code cell below to configure Pandas, create a new `DataFrame` named `df_v` from a ZIP file, and clone the dataset for intermediate checks.

In [5]:
# DO NOT CHANGE THE CODE IN THIS CELL
# Display precision of decimal places to 1
# Precision of decimal places does not have a significance in this case study
pd.set_option('precision', 1)

df_v = pd.read_csv('https://github.com/bdi475/datasets/blob/main/chicago-ridesharing-vehicles.zip?raw=true', compression='zip')

# Used to keep a clean copy
df_v_backup = df_v.copy()

# Display the first 5 rows
df_v.head()

REPORTED_YEAR  REPORTED_MONTH STATE      MAKE       MODEL        COLOR  \
0           2015               3    IL    Toyota  Highlander        Black   
1           2015               3    IL    Toyota       Camry        Black   
2           2015               3    IL    Toyota       Camry        Black   
3           2015               3    IL     Mazda      Mazda3       Silver   
4           2015               3    IL  Chrysler  Pt Cruiser  White/Pearl   

   MODEL_YEAR  NUMBER_OF_TRIPS  MULTIPLE_TNPS  
0        2010              263          False  
1        2008              204          False  
2        2009              122          False  
3        2012              150          False  
4        2008              208          False

---

### 🎯 Part2: Find the number of rows and columns in the dataset

#### 👇 Tasks

- ✔️ Store the number of rows in `df_v` to a new variable named `num_rows`.
- ✔️ Store the number of columns in `df_v` to a new variable named `num_cols`.
- ✔️ Both `num_rows` and `num_cols` must be `int`s.
- ✔️ Use `.shape`, not `len()`.

In [6]:
### BEGIN SOLUTION
num_rows = df_v.shape[0]
num_cols = df_v.shape[1]
### END SOLUTION

print(f'There are {num_rows} rows and {num_cols} columns in the dataset.')
print('🐼 With over a million rows, this dataset is by far the largest one we\'ve worked with so far!!')

There are 1377583 rows and 9 columns in the dataset.
🐼 With over a million rows, this dataset is by far the largest one we've worked with so far!!


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [7]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-02"
    _points = 2

    tc.assertEqual(num_rows, len(df_v_backup.index), f'Number of rows should be {len(df_v_backup.index)}')
    tc.assertEqual(num_cols, len(df_v_backup.columns), f'Number of columns should be {len(df_v_backup.columns)}')
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 3: Concise summary of the dataset

#### 👇 Tasks

- ✔️ Use the `info()` method to print out a concise summary of `df_v`.

#### 🚀 Hints

- `my_dataframe.info()` prints out a concise summary of `my_dataframe`.

In [8]:
### BEGIN SOLUTION
df_v.info()
### END SOLUTION

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1377583 entries, 0 to 1377582
Data columns (total 9 columns):
 #   Column           Non-Null Count    Dtype 
---  ------           --------------    ----- 
 0   REPORTED_YEAR    1377583 non-null  int64 
 1   REPORTED_MONTH   1377583 non-null  int64 
 2   STATE            1377583 non-null  object
 3   MAKE             1377583 non-null  object
 4   MODEL            1377583 non-null  object
 5   COLOR            1377583 non-null  object
 6   MODEL_YEAR       1377583 non-null  int64 
 7   NUMBER_OF_TRIPS  1377583 non-null  int64 
 8   MULTIPLE_TNPS    1377583 non-null  bool  
dtypes: bool(1), int64(4), object(4)
memory usage: 85.4+ MB


In [9]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-03"
    _points = 2
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 4: Total number of trips

#### 👇 Tasks

- ✔️ Calculate the total number of trips (monthly) made by all registered vehicles in `df_v`.
- ✔️ Store the result to a new variable named `total_num_trips`.
- ✔️ `total_num_trips` should be an `int` type (e.g., `int`, `np.int64`).

#### 🚀 Hints

- Sum all values in the `NUMBER_OF_TRIPS` column.

In [10]:
### BEGIN SOLUTION
total_num_trips = df_v['NUMBER_OF_TRIPS'].sum()
### END SOLUTION

print(f'{total_num_trips} trips were made in total.')

328935191 trips were made in total.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [11]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-04"
    _points = 3

    tc.assertEqual(total_num_trips, np.sum(df_v_backup['_'.join(['nUmbEr', 'oF', 'TrIpS']).upper()]))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 5: Average number of trips (monthly)

#### 👇 Tasks

- ✔️ Calculate the average number of trips (monthly) made by all registered vehicles in `df_v`.
- ✔️ Store the result to a new variable named `avg_num_trips`.
- ✔️ `avg_num_trips` should be a `float` type (e.g., `float`, `np.float64`).

#### 🚀 Hints

- Find the mean of all values in the `NUMBER_OF_TRIPS` column.

In [12]:
### BEGIN SOLUTION
avg_num_trips = df_v['NUMBER_OF_TRIPS'].mean()
### END SOLUTION

print(f'On average, a vehicle made {round(avg_num_trips, 1)} trips each month.')

On average, a vehicle made 238.8 trips each month.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [13]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-05"
    _points = 3

    tc.assertAlmostEqual(avg_num_trips, np.mean(df_v_backup['_'.join(['nUmbEr', 'oF', 'TrIpS']).upper()]))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 6: Filter Lincoln vehicles

#### 👇 Tasks

- ✔️ Using `df_v`, filter rows where the vehicle's `MAKE` is `"Lincoln"`.
    - Store the result to a new variable named `df_lincoln`.
- ✔️ `df_v` should remain unaltered after your code.

In [14]:
### BEGIN SOLUTION
df_lincoln = df_v[df_v['MAKE'] == 'Lincoln']
### END SOLUTION

display(df_lincoln.head(5))
print(f'There are {df_lincoln.shape[0]} monthly Lincoln vehicle registrations.')

REPORTED_YEAR  REPORTED_MONTH STATE     MAKE     MODEL  COLOR  \
107           2015               3    IL  Lincoln       Mkz  Black   
468           2015               3    IL  Lincoln       Mks  Black   
731           2015               3    IL  Lincoln       Mks  Black   
768           2015               3    IL  Lincoln  Town Car  Black   
781           2015               3    IL  Lincoln       Mks  Black   

     MODEL_YEAR  NUMBER_OF_TRIPS  MULTIPLE_TNPS  
107        2007              115          False  
468        2009              160          False  
731        2009              152          False  
768        2007              145          False  
781        2010              215          False

There are 24093 monthly Lincoln vehicle registrations.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [15]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-06"
    _points = 2

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_lincoln
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_v_backup.query(f'{"".join(["ma", "ke"]).upper()} == "{"lIncOlN".capitalize()}"')
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  check_like=True)
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 7: Average number of trips (monthly) of Lincoln vehicles

#### 👇 Tasks

- ✔️ Using `df_lincoln`, calculate the average number of trips (monthly) of all Lincoln vehicles.
- ✔️ Store the result to a new variable named `avg_num_trips_lincoln`.
- ✔️ `avg_num_trips_lincoln` should be a `float` type (e.g., `float`, `np.float64`).

#### 🚀 Hints

- Find the mean of all
values in `df_lincoln`'s `NUMBER_OF_TRIPS` column.

In [16]:
### BEGIN SOLUTION
avg_num_trips_lincoln = df_lincoln['NUMBER_OF_TRIPS'].mean()
### END SOLUTION

print(f'Lincoln drivers made {round(avg_num_trips_lincoln, 1)} trips (monthly) on average.')
print(f'This is higher than the overall average of {round(avg_num_trips, 1)} trips (monthly).')

Lincoln drivers made 271.3 trips (monthly) on average.
This is higher than the overall average of 238.8 trips (monthly).


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [17]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-07"
    _points = 3

    tc.assertAlmostEqual(avg_num_trips_lincoln, df_v_backup.query(f'{"".join(["mA", "Ke"]).upper()} == "{"LiNCoLN".capitalize()}"') \
        ["_".join(['NuMBeR', 'oF', 'tRiPs']).upper()].mean())
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 8: Red Tesla Vehicles

#### 👇 Tasks

- ✔️ Using `df_v`, filter rows where:
    - `MAKE` is `Tesla`, **AND**
    - `COLOR` is `Red`.
- ✔️ Store the result to a new variable named `df_red_tesla`.
- ✔️ `df_v` should remain unaltered after your code.

#### 🚀 Hints

- Use the logical AND operator (`&`) to check whether two conditions are both satisfied.

In [18]:
### BEGIN SOLUTION
df_red_tesla = df_v[(df_v['MAKE'] == 'Tesla') & (df_v['COLOR'] == 'Red')]
### END SOLUTION

df_red_tesla.head(5)

REPORTED_YEAR  REPORTED_MONTH STATE   MAKE    MODEL COLOR  MODEL_YEAR  \
660034           2018               6    IL  Tesla  Model S   Red        2013   
798621           2018              10    IL  Tesla  Model 3   Red        2018   
808521           2018              11    IL  Tesla  Model S   Red        2013   
816799           2018              11    IL  Tesla  Model 3   Red        2018   
938426           2019               3    IL  Tesla  Model 3   Red        2018   

        NUMBER_OF_TRIPS  MULTIPLE_TNPS  
660034              143           True  
798621              184          False  
808521              305           True  
816799              142          False  
938426              220           True

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [19]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-08"
    _points = 4

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_v_backup
                                      .query(f'{"".join(["mA", "kE"]).upper()} == "{"".join(["T", "e", "s", "l", "a"])}" \
                                              & {"".join(["c", "OLO", "r"]).upper()} == "{("r" + "E" + "D").capitalize()}"')
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_red_tesla
                                      .sort_values(df_red_tesla.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 9: Porsche + Hummer

#### 👇 Tasks

- ✔️ Using `df_v`, filter rows where:
    - `MAKE` is `Porsche`, **OR**
    - `MAKE` is `Hummer`.
- ✔️ Store the result to a new variable named `df_porsche_hummer`.
- ✔️ `df_v` should remain unaltered after your code.

#### 🚀 Hints

- Use the logical OR operator (`|`) to check whether at least one of the two conditions.

In [20]:
### BEGIN SOLUTION
df_porsche_hummer = df_v[(df_v['MAKE'] == 'Porsche') | (df_v['MAKE'] == 'Hummer')]
### END SOLUTION

display(df_porsche_hummer.head(5))
print(f'There are {df_porsche_hummer.shape[0]} Porsche and Hummer vehicles.')

REPORTED_YEAR  REPORTED_MONTH STATE     MAKE    MODEL     COLOR  \
1882            2015               3    IL   Hummer       H3     Black   
13550           2015               5    IL   Hummer       H3  "Silver"   
22851           2015               6    IL  Porsche  Cayenne     Black   
27740           2015               7    IL   Hummer       H3     Black   
30697           2015               8    IL   Hummer       H3     Black   

       MODEL_YEAR  NUMBER_OF_TRIPS  MULTIPLE_TNPS  
1882         2008              111          False  
13550        2006              140          False  
22851        2005              113          False  
27740        2007              139          False  
30697        2006              119          False

There are 411 Porsche and Hummer vehicles.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [21]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-09"
    _points = 4

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_v_backup
                                      .query(f'{"mAKe".upper()} == "{"PORsCHE".capitalize()}" \
                                          | {"mAke".upper()} == "{"HuMMeR".capitalize()}"')
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_porsche_hummer
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 10: Maximum number of trips

#### 👇 Tasks

- ✔️ Find the maximum number of monthly trips for a given vehicle in `df_v`.
- ✔️ Store the value in a new variable named `max_num_trips`.

#### 🚀 Hints

- Find the maximum value in the `"NUMBER_OF_TRIPS"` column.

In [22]:
### BEGIN SOLUTION
max_num_trips = df_v['NUMBER_OF_TRIPS'].max()
### END SOLUTION

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [23]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-10"
    _points = 3

    tc.assertEqual(max_num_trips, 999)
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 11: Black Lincoln Mkzs with 999+ trips

#### 👇 Tasks

- ✔️ Using `df_v`, filter rows where:
    - `MODEL` is `Mkz`, **AND**
    - `COLOR` is `Black`, **AND**
    - `NUMBER_OF_TRIPS` is greater than or equal to `999`.
- ✔️ Store the result to a new variable named `df_black_mkz`.
- ✔️ `df_v` should remain unaltered after your code.

#### 🚀 Hints

- Use the logical AND operator (`&`) with parentheses to check whether two or more conditions are satisfied.
- **Example**: `df_filtered = df[(condition1) & (condition2) & (condition3)]`

#### 🔑 Expected Output

Your index column may contain different values.

|         |   REPORTED_YEAR |   REPORTED_MONTH | STATE   | MAKE    | MODEL   | COLOR   |   MODEL_YEAR |   NUMBER_OF_TRIPS | MULTIPLE_TNPS   |
|--------:|----------------:|-----------------:|:--------|:--------|:--------|:--------|-------------:|------------------:|:----------------|
|  782967 |            2018 |                1 | IL      | Lincoln | Mkz     | Black   |         2017 |               999 | True            |
| 1064108 |            2019 |                3 | IL      | Lincoln | Mkz     | Black   |         2016 |               999 | True            |
| 1228454 |            2019 |                8 | IL      | Lincoln | Mkz     | Black   |         2018 |               999 | True            |
| 1337522 |            2020 |                2 | IL      | Lincoln | Mkz     | Black   |         2017 |               999 | True            |

In [24]:
### BEGIN SOLUTION
df_black_mkz = df_v[(df_v['MODEL'] == 'Mkz') & (df_v['COLOR'] == 'Black') & (df_v['NUMBER_OF_TRIPS'] >= 999)]
### END SOLUTION

display(df_black_mkz)
print(f'There are {df_black_mkz.shape[0]} black Mkzs that have made 999+ trips in a given month.')

REPORTED_YEAR  REPORTED_MONTH STATE     MAKE MODEL  COLOR  \
539360            2018               1    IL  Lincoln   Mkz  Black   
946218            2019               3    IL  Lincoln   Mkz  Black   
1105919           2019               8    IL  Lincoln   Mkz  Black   
1287915           2020               2    IL  Lincoln   Mkz  Black   

         MODEL_YEAR  NUMBER_OF_TRIPS  MULTIPLE_TNPS  
539360         2017              999           True  
946218         2016              999           True  
1105919        2018              999           True  
1287915        2017              999           True

There are 4 black Mkzs that have made 999+ trips in a given month.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [25]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-11"
    _points = 4

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_v_backup
                                      .query(f'{"mOdEL".upper()} == "{"mKz".capitalize()}" \
                                          & {"cOlOr".upper()} == "{"bLaCK".capitalize()}" \
                                          & {"_".join(["nUMBEr", "OF", "TripS"]).upper()} > (100 * 10 - 2)')
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_black_mkz
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 12: Latest model year vehicles

#### 👇 Tasks

- ✔️ Using `df_v`, filter rows where the vehicle's `MODEL_YEAR` (NOT `REPORTED_YEAR`) is 2022 or newer.
- ✔️ Store the result to a new variable named `df_latest_models`.
- ✔️ `df_v` should remain unaltered after your code.

#### 🔑 Expected Output

Your index column may contain different values.

|         |   REPORTED_YEAR |   REPORTED_MONTH | STATE   | MAKE    | MODEL    | COLOR   |   MODEL_YEAR |   NUMBER_OF_TRIPS | MULTIPLE_TNPS   |
|--------:|----------------:|-----------------:|:--------|:--------|:---------|:--------|-------------:|------------------:|:----------------|
| 1599787 |            2021 |                3 | IL      | Acura   | Mdx      | Black   |         2022 |               122 | False           |
| 1610215 |            2021 |                4 | IL      | Acura   | Mdx      | Black   |         2022 |               148 | True            |
| 1623251 |            2021 |                5 | IL      | Kia     | Carnival | Black   |         2022 |               123 | False           |
| 1631454 |            2021 |                6 | IL      | Hyundai | Tucson   | Blue    |         2022 |               121 | True            |
| 1639844 |            2021 |                6 | IL      | Kia     | Carnival | Black   |         2022 |               449 | False           |

In [26]:
### BEGIN SOLUTION
df_latest_models = df_v[df_v['MODEL_YEAR'] >= 2022]
### END SOLUTION

display(df_latest_models.head(5))
print(f'There are {df_latest_models.shape[0]} latest model registrations.')

Empty DataFrame
Columns: [REPORTED_YEAR, REPORTED_MONTH, STATE, MAKE, MODEL, COLOR, MODEL_YEAR, NUMBER_OF_TRIPS, MULTIPLE_TNPS]
Index: []

There are 0 latest model registrations.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [27]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-12"
    _points = 3

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_v_backup
                                      .query(f'{"_".join(["MoDeL", "YeaR"]).upper()} > (1010 * 2 + 1)')
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_latest_models
                                      .sort_values(df_v_backup.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 13: Compare average number of trips by `MULTIPLE_TNPS`

A driver can choose to drive for only one ridesharing service (e.g., only for Uber) or choose to drive for two or more services (e.g., for both Uber **and** Lyft).

`MULTIPLE_TNPS` column indicates whether a driver is registered for multiple rideshare services. The `MULTIPLE_TNPS` value of a registered vehicle will be:
- `False` if the driver is only registered for one service (e.g., only for Uber)
- `True` if the driver is registered for two or more services (e.g., Uber **and** Lyft)

**Question**: Do drivers make more trips on average if they are registered in multiple rideshare services?

Your goal is to answer that question using data. 📐

#### 👇 Tasks

- ✔️ Using `df_v`, calculate the following two metrics:
    1. Average number of trips by vehicles where `MULTIPLE_TNPS` is `False`.
        - ✏️ Store this to a new variable named `avg_num_trips_single_tnps`.
        - `avg_num_trips_single_tnps` should be a `float`-typed variable.
    2. Average number of trips by vehicles where `MULTIPLE_TNPS` is `True`.
        - ✏️ Store this to a new variable named `avg_num_trips_multiple_tnps`.
        - `avg_num_trips_multiple_tnps` should be a `float`-typed variable.
- ✔️ `df_v` should remain unaltered after your code.

#### 🚀 Hints

- There are many ways to do this. The process described below is only one of them.
- Create the following two `DataFrame`s using filter.
    1. `df_single_tnps`: Only contains rows where `MULTIPLE_TNPS` is `False`.
    2. `df_multiple_tnps`: Only contains rows where `MULTIPLE_TNPS` is `True`.
- For each of the two `DataFrame`s you've created, calculate the average of `NUMBER_OF_TRIPS`.
    - **Example**: `df_single_tnps['NUMBER_OF_TRIPS'].mean()`

In [28]:
### BEGIN SOLUTION
df_single_tnps = df_v[~df_v['MULTIPLE_TNPS']]
df_multiple_tnps = df_v[df_v['MULTIPLE_TNPS']]

avg_num_trips_single_tnps = df_single_tnps['NUMBER_OF_TRIPS'].mean()
avg_num_trips_multiple_tnps = df_multiple_tnps['NUMBER_OF_TRIPS'].mean()
### END SOLUTION

print(f'Drivers driving for a single rideshare service made {round(avg_num_trips_single_tnps, 1)} trips (monthly) on average.')
print(f'Drivers driving for multiple rideshare services made {round(avg_num_trips_multiple_tnps, 1)} trips (monthly) on average.')

Drivers driving for a single rideshare service made 223.0 trips (monthly) on average.
Drivers driving for multiple rideshare services made 286.0 trips (monthly) on average.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [29]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-13"
    _points = 6

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    col_check = '_'.join(['MulTiPle', ''.join(['T', 'n', 'P', 's'])]).upper()
    col_metric = '_'.join([''.join(['NuM', 'beR']), 'oF', ''.join(['T', 'r', 'I', 'P', 's'])]).upper()

    tc.assertAlmostEqual(avg_num_trips_single_tnps, np.mean(df_v.query(f'~{col_check}')[col_metric]))
    tc.assertAlmostEqual(avg_num_trips_multiple_tnps, np.mean(df_v.query(f'{col_check}')[col_metric]))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

## 🔮 Make/Model/Color Analysis

So far, we've looked at *individual* vehicle registrations. In this section, we will *aggregate* vehicles by one or more criteria.

---

### 🎯 Part 14: Create a Summary of Metrics by Vehicle Manufacturer

Our first *level of aggregation* is the vehicle manufacturer (`MAKE`). For each vehicle manufacturer, you'll be calculating the total number of registered vehicles (`NUM_REG`, total number of trips (`TOTAL_NUM_TRIPS`), and the average number of trips (`AVG_NUM_TRIPS`).

#### 👇 Tasks

- ✔️ Copy the code in the image above to the code cell below.
    - 😺 *Try* to understand what the code does.
- ✔️ `df_v` should remain unaltered after your code.
- ✔️ **We will provide you the code for this deliverable.**

![Groupby make](https://github.com/bdi475/images/blob/main/case-studies/rideshare-vehicles/carbon-groupby-make-01.png?raw=true)

In [30]:
### BEGIN SOLUTION
# Group by vehicle manufacturer
df_by_make = df_v.groupby('MAKE', as_index=False).agg({
    'NUMBER_OF_TRIPS': ['count', 'sum', 'mean']
})

# Reset columns
df_by_make.columns = ['MAKE', 'NUM_REG', 'TOTAL_NUM_TRIPS', 'AVG_NUM_TRIPS']

# Display first 6 rows
df_by_make.head(6)
### END SOLUTION

MAKE  NUM_REG  TOTAL_NUM_TRIPS  AVG_NUM_TRIPS
0       Acura    19353          4326062          223.5
1  Alfa Romeo       30             7243          241.4
2  Am General       31             6889          222.2
3        Audi     4857          1026884          211.4
4         Bmw     8025          1619829          201.8
5       Buick    15234          3164821          207.7

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [31]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-14"
    _points = 3

    df_check = df_v_backup.groupby(''.join(['m', 'ak', 'e']).upper()) \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'ou', 'n', 't']), np.sum, np.mean]}) \
        .reset_index()
    df_check.columns = ['AVG_NUM_TRIPS', 'TOTAL_NUM_TRIPS', 'NUM_REG', 'MAKE'][::-1]
    df_check = df_check.sort_values(df_check.columns.tolist()).reset_index(drop=True)
    df_by_make_backup = df_check.copy()

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_check,
                                  df_by_make
                                      .sort_values(df_by_make.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 15: Sort by number of registered vehicles

Can you sort `df_by_make` by popularity (i.e., number of registrations)?

#### 👇 Tasks

- ✔️ Sort `df_by_make` by `NUM_REG` in **descending** order.
- ✔️ Store the sorted result to a new variable named `df_by_make_popularity`.
- ✔️ `df_by_make` should remain unaltered after your code (out-of-place sort).
    - In other words, your code should store the sorted result to `df_by_make_popularity` **without affecting `df_by_make`**.

#### 🔑 Expected Output

Your index column may contain different values.

|    | MAKE      |   NUM_REG |   TOTAL_NUM_TRIPS | AVG_NUM_TRIPS |
|---:|:----------|----------:|------------------:|--------------:|
| 69 | Toyota    |    443236 |         117885589 |         266.0 |
| 54 | Nissan    |    211955 |          50103680 |         236.4 |
| 26 | Honda     |    148422 |          35092978 |         236.4 |
| 13 | Chevrolet |    140633 |          30354594 |         215.8 |
| 28 | Hyundai   |    130522 |          31992661 |         245.1 |

In [32]:
### BEGIN SOLUTION
df_by_make_popularity = df_by_make.sort_values('NUM_REG', ascending=False)
### END SOLUTION

df_by_make_popularity.head(5)

MAKE  NUM_REG  TOTAL_NUM_TRIPS  AVG_NUM_TRIPS
48     Toyota   351368         92779589          264.1
36     Nissan   178602         42355093          237.1
16      Honda   121011         28441079          235.0
8   Chevrolet   118649         25589661          215.7
18    Hyundai   109433         26840858          245.3

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [33]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-15"
    _points = 3

    df_check = df_by_make_backup \
        .sort_values('_'.join(['num', 'reg']).upper(), ascending=bool(0), inplace=bool(0))

    pd.testing.assert_frame_equal(df_by_make_backup
                                      .sort_values(df_by_make_backup.columns.tolist())
                                      .reset_index(drop=True),
                                  df_by_make
                                      .sort_values(df_by_make.columns.tolist())
                                      .reset_index(drop=True),
                                 'df_by_make should remain unaltered.')

    pd.testing.assert_series_equal(df_check['NUM_REG'].reset_index(drop=True),
                                  df_by_make_popularity['NUM_REG'].reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 16: Sort by average number of trips (monthly)

#### 👇 Tasks

- ✔️ Sort `df_by_make` from 🎯 Part 14 (**NOT** `df_by_make_popularity`) by `AVG_NUM_TRIPS` in **descending** order.
- ✔️ Store the sorted result to a new variable named `df_by_make_trips`.
- ✔️ Drop `NUM_REG` and `TOTAL_NUM_TRIPS` columns from `df_by_make_trips`.
    - After your code runs, `df_by_make_trips` should only have two columns - `MAKE` and `AVG_NUM_TRIPS`.
- ✔️ `df_by_make` should remain unaltered after your code.
    - In other words, your code should (1) store the sorted result to `df_by_make_trips` and (2) drop `NUM_REG` and `TOTAL_NUM_TRIPS` columns **without affecting `df_by_make`**.
- ⚠️ **Warning**: Running your code more than once may throw an error since you cannot drop a column that has already been dropped. Despite the error message, this is not a huge problem as it won't have any effect on your `DataFrame`. The error message will go away if you go back a few steps and restart with a *fresh* `df_by_make`.

#### 🔑 Expected Output

Your index column may contain different values.

|    | MAKE           |   AVG_NUM_TRIPS |
|---:|:---------------|----------------:|
| 67 | Taxi           |           499.5 |
| 72 | Unique         |           483.0 |
| 14 | Chevrolt       |           446.0 |
| 40 | Lincoin        |           428.5 |
| 22 | General Motors |           396.0 |

In [34]:
### BEGIN SOLUTION
df_by_make_trips = df_by_make.sort_values('AVG_NUM_TRIPS', ascending=False)
df_by_make_trips.drop(columns=['NUM_REG', 'TOTAL_NUM_TRIPS'], inplace=True)
### END SOLUTION

df_by_make_trips.head(5)

MAKE  AVG_NUM_TRIPS
6         Byd          288.1
26    Lincoln          271.3
28   Mercedes          269.5
48     Toyota          264.1
35  Mv Agusta          261.0

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [35]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-16"
    _points = 5

    df_check = df_by_make_backup \
            .sort_values('_'.join(['avg', 'num', 'trips']).upper(), ascending=bool(0), inplace=bool(0)) \
            .drop(columns=['NUM_REG', 'TOTAL_NUM_TRIPS'])

    pd.testing.assert_frame_equal(df_check.reset_index(drop=True),
                                  df_by_make_trips.reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 17: Average number of trips of Japanese Big 3

#### 👇 Tasks

- ✔️ Using `df_by_make_trips`, retrieve the following three values.
    1. `avg_num_trips_toyota`: Average number of trips (monthly) of Toyota vehicles
    2. `avg_num_trips_nissan`: Average number of trips (monthly) of Nissan vehicles
    3. `avg_num_trips_honda`: Average number of trips (monthly) of Honda vehicles
- ✔️ All 3 variables must be `float`-typed.
- ✔️ `df_by_make_trips` should remain unaltered after your code.
- ⚠️ You must perform this task **programmatically**. You cannot manually enter the numbers (e.g., `avg_num_trips_toyota = 264.1` is NOT a correct answer).

#### 🚀 Hints

To retrieve the value of `major1` column of `Dylan` from `df_you`, you can use the following code:

```python
df_you[df_you['name'] == 'Dylan']['major1'].iloc[0]
```

**The code above will...**

1. Filter row(s) where `name` is `Dylan`.
2. Retrieve the `major1` column as a Pandas `Series`.
3. Get the value of the first row in the `Series` (`iloc[0]` returns the value of the first row).

In [36]:
### BEGIN SOLUTION
avg_num_trips_toyota = df_by_make_trips[df_by_make_trips['MAKE'] == 'Toyota']['AVG_NUM_TRIPS'].iloc[0]
avg_num_trips_nissan = df_by_make_trips[df_by_make_trips['MAKE'] == 'Nissan']['AVG_NUM_TRIPS'].iloc[0]
avg_num_trips_honda = df_by_make_trips[df_by_make_trips['MAKE'] == 'Honda']['AVG_NUM_TRIPS'].iloc[0]
### END SOLUTION

print(f'Toyota vehicles have an average number of {round(avg_num_trips_toyota, 1)} trips (monthly).')
print(f'Nissan vehicles have an average number of {round(avg_num_trips_nissan, 1)} trips (monthly).')
print(f'Honda vehicles have an average number of {round(avg_num_trips_honda, 1)} trips (monthly).')

Toyota vehicles have an average number of 264.1 trips (monthly).
Nissan vehicles have an average number of 237.1 trips (monthly).
Honda vehicles have an average number of 235.0 trips (monthly).


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [37]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-17"
    _points = 6

    df_check = df_v_backup.groupby('MAKE') \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'ou', 'n', 't']), np.sum, np.mean]}) \
        .reset_index()
    df_check.columns = ['AVG_NUM_TRIPS', 'TOTAL_NUM_TRIPS', 'NUM_REG', 'MAKE'][::-1]

    col_check = '_'.join(['avg', 'num', 'trips']).upper()
    make_check = list(map(lambda x: x.capitalize(), ['toYOta', 'niSSan', 'hONDa']))
    var_check = [avg_num_trips_toyota, avg_num_trips_nissan, avg_num_trips_honda]

    for i in range(len(var_check)):
        tc.assertAlmostEqual(var_check[i], df_check.query(f'MAKE == "{make_check[i]}"')[col_check].iloc[0])
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 18: Create a Summary of Metrics by Vehicle Model

Our second *level of aggregation* is the vehicle manufacturer (`MAKE`) **and** model (`MODEL`). For each vehicle model, you'll be calculating the total number of registered vehicles (`NUM_REG`, total number of trips (`TOTAL_NUM_TRIPS`), and the average number of trips (`AVG_NUM_TRIPS`).

**Again, we will provide you the code for this deliverable.**

![Groupby make](https://github.com/bdi475/images/blob/main/case-studies/rideshare-vehicles/carbon-groupby-make-and-model-01.png?raw=true)

#### 👇 Tasks

- ✔️ Copy the code in the image above to the code cell below.
    - 😺 *Try* to understand what the code does.
- ✔️ `df_v` should remain unaltered after your code.

In [38]:
### BEGIN SOLUTION
# Group by MAKE & MODEL and calculate stats
df_by_model = df_v.groupby(['MAKE', 'MODEL'], as_index=False).agg({
    'NUMBER_OF_TRIPS': ['count', 'sum', 'mean']
})

# Rename columns
df_by_model.columns = ['MAKE', 'MODEL', 'NUM_REG', 'TOTAL_NUM_TRIPS', 'AVG_NUM_TRIPS']

# Display first 6 rows
df_by_model.head(6)
### END SOLUTION

MAKE MODEL  NUM_REG  TOTAL_NUM_TRIPS  AVG_NUM_TRIPS
0  Acura   Ilx      826           180669          218.7
1  Acura   Mdx     9811          2235004          227.8
2  Acura   Rdx     1454           326177          224.3
3  Acura    Rl      329            65065          197.8
4  Acura   Rlx       36             6447          179.1
5  Acura    Tl     3367           733591          217.9

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [39]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-18"
    _points = 3

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'ou', 'n', 't']), np.sum, np.mean]}) \
        .reset_index()
    df_check.columns = ['AVG_NUM_TRIPS', 'TOTAL_NUM_TRIPS', 'NUM_REG', 'MODEL', 'MAKE'][::-1]

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_check
                                      .sort_values(df_check.columns.tolist())
                                      .reset_index(drop=True),
                                  df_by_model
                                      .sort_values(df_by_model.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 19: Rename columns

#### 👇 Tasks

- ✔️ Rename the following columns in `df_by_model`:
    1. `MAKE` to `manufacturer`
    2. `MODEL` to `model`
    3. `NUM_REG` to `registrations`
    4. `AVG_NUM_TRIPS` to `avg_trips`
- ⚠️ Don't worry about the `TOTAL_NUM_TRIPS` column. You'll drop this column soon.

#### 🚀 Hints

To rename a column named `col1` to `col1_renamed` and `col2` to `col2_renamed` in `my_dataframe`, use the following code.

![rename inplace](https://github.com/bdi475/images/blob/main/case-studies/rideshare-vehicles/carbon-rename-columns-inplace-02.png?raw=true)

#### 🔑 Expected Output

Your index column may contain different values.

|    | manufacturer   | model   |   registrations |   TOTAL_NUM_TRIPS |   avg_trips |
|---:|:---------------|:--------|----------------:|------------------:|------------:|
|  0 | Acura          | 3.2 Tl  |               1 |               313 |     313     |
|  1 | Acura          | Aw109   |               4 |               553 |     138.25  |
|  2 | Acura          | Camry   |              13 |              2323 |     178.692 |
|  3 | Acura          | Cl      |               5 |              1461 |     292.2   |
|  4 | Acura          | Csx     |               5 |              1387 |     277.4   |

In [40]:
### BEGIN SOLUTION
df_by_model.rename(columns={
    'MAKE': 'manufacturer', 
    'MODEL': 'model',
    'NUM_REG': 'registrations',
    'AVG_NUM_TRIPS': 'avg_trips'
}, inplace=True)
### END SOLUTION

df_by_model.head(5)

manufacturer model  registrations  TOTAL_NUM_TRIPS  avg_trips
0        Acura   Ilx            826           180669      218.7
1        Acura   Mdx           9811          2235004      227.8
2        Acura   Rdx           1454           326177      224.3
3        Acura    Rl            329            65065      197.8
4        Acura   Rlx             36             6447      179.1

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [41]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-19"
    _points = 3

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'ou', 'n', 't']), np.sum, np.mean]}) \
        .reset_index()
    df_check.columns = ['avg_trips', 'TOTAL_NUM_TRIPS', 'registrations', 'model', 'manufacturer'][::-1]

    pd.testing.assert_frame_equal(df_check
                                      .sort_values(df_check.columns.tolist())
                                      .reset_index(drop=True),
                                  df_by_model
                                      .sort_values(df_by_model.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 20: Drop `TOTAL_NUM_TRIPS` column

#### 👇 Tasks

- ✔️ Drop the `TOTAL_NUM_TRIPS` column from `df_by_model`.
- ⚠️ **Warning**: Running your code more than once may throw an error since you cannot drop a column that has already been dropped. Despite the error message, this is not a huge problem as it won't have any effect on your `DataFrame`. The error message will go away if you go back a few steps and restart with a *fresh* `df_by_model`.

#### 🔑 Expected Output

Your index column may contain different values.

|    | manufacturer   | model   |   registrations |   avg_trips |
|---:|:---------------|:--------|----------------:|------------:|
|  0 | Acura          | 3.2 Tl  |               1 |     313     |
|  1 | Acura          | Aw109   |               4 |     138.25  |
|  2 | Acura          | Camry   |              13 |     178.692 |
|  3 | Acura          | Cl      |               5 |     292.2   |
|  4 | Acura          | Csx     |               5 |     277.4   |

In [42]:
### BEGIN SOLUTION
df_by_model.drop(columns=['TOTAL_NUM_TRIPS'], inplace=True)
### END SOLUTION

df_by_model.head(5)

manufacturer model  registrations  avg_trips
0        Acura   Ilx            826      218.7
1        Acura   Mdx           9811      227.8
2        Acura   Rdx           1454      224.3
3        Acura    Rl            329      197.8
4        Acura   Rlx             36      179.1

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [43]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-20"
    _points = 2

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'ou', 'n', 't']), np.mean]}) \
        .reset_index()
    df_check.columns = list(map(lambda x: x.lower(), ['AVg_TriPS', 'ReGisTraTiONS', 'MoDeL', 'ManUfacTurer'][::-1]))

    pd.testing.assert_frame_equal(df_check
                                      .sort_values(df_check.columns.tolist())
                                      .reset_index(drop=True),
                                  df_by_model
                                      .sort_values(df_by_model.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 21: Sort by number of registered vehicles

Can you sort `df_by_model` by popularity?

#### 👇 Tasks

- ✔️ Sort `df_by_model` by `registrations` in **descending** order.
- ✔️ Update `df_by_model` without creating a new variable.
    - In other words, you're performing an **in-place** sort.

#### 🔑 Expected Output

Your index column may contain different values.

|      | manufacturer   | model   |   registrations |   avg_trips |
|-----:|:---------------|:--------|----------------:|------------:|
| 1170 | Toyota         | Camry   |          157372 |     270.232 |
| 1179 | Toyota         | Corolla |           79926 |     259.916 |
|  973 | Nissan         | Altima  |           71884 |     238.125 |
| 1211 | Toyota         | Prius   |           71005 |     280.718 |
| 1014 | Nissan         | Sentra  |           65132 |     244.764 |

In [44]:
### BEGIN SOLUTION
df_by_model.sort_values('registrations', ascending=False, inplace=True)
### END SOLUTION

df_by_model.head(5)

manufacturer    model  registrations  avg_trips
589       Toyota    Camry         129143      269.6
594       Toyota  Corolla          64887      259.6
493       Nissan   Altima          61352      239.7
607       Toyota    Prius          56870      278.4
509       Nissan   Sentra          54599      245.6

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [45]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-21"
    _points = 2

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): [''.join(['c', 'o', 'u', 'n', 't']), np.mean]}) \
        .sort_values(('NUMBER_OF_TRIPS', 'count'), ascending=False) \
        .reset_index()
    df_check.columns = list(map(lambda x: x.lower(), ['AVg_TriPS', 'ReGisTraTiONS', 'MoDeL', 'ManUfacTurer'][::-1]))

    pd.testing.assert_series_equal(df_check['registrations'].reset_index(drop=True),
                                   df_by_model['registrations'].reset_index(drop=True))
    pd.testing.assert_frame_equal(df_check
                                      .sort_values(['registrations', 'manufacturer', 'model'], ascending=[False, True, True])
                                      .reset_index(drop=True),
                                  df_by_model
                                      .sort_values(['registrations', 'manufacturer', 'model'], ascending=[False, True, True])
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 22: Number of Lincoln Ls registrations

#### 👇 Tasks

- ✔️ Using `df_by_model`, retrieve the number of Lincoln LS registrations.
    - `manufacturer == 'Lincoln'`
    - `model == 'Ls'` (case-sensitive)
- ✔️ Store the result to a new variable named `num_lincoln_ls_registrations`.
- ✔️ `num_lincoln_ls_registrations` should be an `int` type.
- ✔️ `df_by_model` should remain unaltered after your code.
- ⚠️ Filtering by the `model` column only (`model == 'Ls'`) may yield incorrect result as Lexus also has a model with the same name (Lexus LS). Use both `manufacturer` and `model` columns to find the matching row.

#### 🚀 Hints

To retrieve the `fav_restaurant` column value of a person where `name == 'Mark'` **and** `city == 'Metamora'`, use the code below.

**The code below will...**

1. Filter row(s) where `name` is `Josh` **and** `city` is `Chicago`.
2. Retrieve the `fav_restaurant` column as a Pandas `Series`.
3. Get the value of the first element in the `Series` (`iloc[0]` returns the value of the element at index position 0, which is the first element).

```python
df_you[(df_you['name'] == 'Josh') & (df_you['city'] == 'Chicago')]['fav_restaurant'].iloc[0]
```

In [46]:
### BEGIN SOLUTION
num_lincoln_ls_registrations = df_by_model[(df_by_model['manufacturer'] == 'Lincoln') & (df_by_model['model'] == 'Ls')]['registrations'].iloc[0]
### END SOLUTION

print(f'There are {num_lincoln_ls_registrations} Lincoln Ls registrations.')

There are 92 Lincoln Ls registrations.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [47]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-22"
    _points = 2

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): ''.join(['c', 'o', 'u', 'n', 't'])}) \
        .reset_index()

    tc.assertEqual(num_lincoln_ls_registrations,
                   df_check.query(f'MAKE == "{"lInCoLn".capitalize()}" & MODEL == "{"lS".capitalize()}"') \
                       ['_'.join(['nUmBer', 'oF', 'tRiPs']).upper()].iloc[0])
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 23: Number of Ford 500 registrations

This part is very similar to the previous one.

#### 👇 Tasks

- ✔️ Using `df_by_model`, retrieve the number of Ford 500 registrations.
    - `manufacturer == 'Ford'`
    - `model == '500'` (⚠️ string `'500'`, not number `500`)
- ✔️ Store the result to a new variable named `num_ford_500_registrations`.
- ✔️ `num_ford_500_registrations` should be an `int` type.
- ✔️ `df_by_model` should remain unaltered after your code.
- ⚠️ Filtering by the `model` column only (`model == '500'`) may yield incorrect result as Fiat also has a model with the same name (Fiat 500). Use both `manufacturer` and `model` columns to find the matching row.

In [48]:
### BEGIN SOLUTION
num_ford_500_registrations = df_by_model[(df_by_model['manufacturer'] == 'Ford') & (df_by_model['model'] == '500')]['registrations'].iloc[0]
### END SOLUTION

print(f'There are {num_ford_500_registrations} Ford 500 registrations.')

There are 94 Ford 500 registrations.


#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [49]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-23"
    _points = 2

    df_check = df_v_backup.groupby(['MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): ''.join(['c', 'o', 'u', 'n', 't'])}) \
        .reset_index()

    tc.assertEqual(num_ford_500_registrations,
                   df_check.query(f'MAKE == "{"fOrD".capitalize()}" & MODEL == "{5 * 5 * 4 * 5}"') \
                       ['_'.join(['nUmBer', 'oF', 'tRiPs']).upper()].iloc[0])
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 24: Create a Summary of Metrics by Vehicle Model and Color

Our final *level of aggregation* is the vehicle manufacturer (`MAKE`), model (`MODEL`), and color (`COLOR`).

#### 👇 Tasks

- ✔️ Copy the code in the image above to the code cell below.
    - 😺 *Try* to understand what the code does.
- ✔️ `df_v` should remain unaltered after your code.
- ✔️  **Again, we will provide you the code for this deliverable.**

![Groupby make](https://github.com/bdi475/images/blob/main/case-studies/rideshare-vehicles/carbon-groupby-make-model-color-01.png?raw=true)

#### 🔑 Expected Output

Your index column may contain different values.

|    | MAKE   | MODEL   | COLOR       |   REPORTED_MONTH |
|---:|:-------|:--------|:------------|-----------------:|
|  0 | Acura  | 3.2 Tl  | White/Pearl |                1 |
|  1 | Acura  | Aw109   | White       |                4 |
|  2 | Acura  | Camry   | Black       |                9 |
|  3 | Acura  | Camry   | White       |                4 |
|  4 | Acura  | Cl      | Gray        |                3 |

In [50]:
### BEGIN SOLUTION
# Group by MAKE/MODEL/COLOR and count rows
df_by_color = df_v.groupby(['MAKE', 'MODEL', 'COLOR'], as_index=False).agg({
    'REPORTED_MONTH': 'count'
})

# Display first 6 rows
df_by_color.head(5)
### END SOLUTION

MAKE MODEL   COLOR  REPORTED_MONTH
0  Acura   Ilx       "               1
1  Acura   Ilx  "Grey"               1
2  Acura   Ilx   Black             410
3  Acura   Ilx    Blue              35
4  Acura   Ilx   Brown               7

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [51]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-24"
    _points = 2

    df_check = df_v_backup.groupby(['COLOR', 'MODEL', 'MAKE'][::-1]) \
        .agg({'nUMbEr_of_trIPs'.upper(): ''.join(['c', 'ou', 'n', 't'])}) \
        .reset_index()
    df_check.columns = ['REPORTED_MONTH', 'COLOR', 'MODEL', 'MAKE'][::-1]

    tc.assertEqual(df_v.shape, df_v_backup.shape, 'df_v should remain unaltered.')

    pd.testing.assert_frame_equal(df_check
                                      .sort_values(df_check.columns.tolist())
                                      .reset_index(drop=True),
                                  df_by_color
                                      .sort_values(df_by_color.columns.tolist())
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 25: Rename columns

#### 👇 Tasks

- ✔️ Rename the following columns in `df_by_color` **in-place**:
    1. `MAKE` to `manufacturer`
    2. `MODEL` to `model`
    3. `COLOR` to `color`
    4. `REPORTED_MONTH` to `registrations`

#### 🔑 Expected Output

Your index column may contain different values.
    
|    | manufacturer   | model   | color       |   registrations |
|---:|:---------------|:--------|:------------|----------------:|
|  0 | Acura          | 3.2 Tl  | White/Pearl |               1 |
|  1 | Acura          | Aw109   | White       |               4 |
|  2 | Acura          | Camry   | Black       |               9 |

In [52]:
### BEGIN SOLUTION
df_by_color.rename(columns={
    'MAKE': 'manufacturer', 
    'MODEL': 'model',
    'COLOR': 'color',
    'REPORTED_MONTH': 'registrations'
}, inplace=True)
### END SOLUTION

df_by_color.head(3)

manufacturer model   color  registrations
0        Acura   Ilx       "              1
1        Acura   Ilx  "Grey"              1
2        Acura   Ilx   Black            410

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [53]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-25"
    _points = 3

    df_check = df_v_backup.groupby(['COLOR', 'MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): ''.join(['c', 'ou', 'n', 't'])}) \
        .reset_index()
    df_check.columns = list(map(lambda x: x.lower(), ['RegIstraTioNs', 'CoLoR', 'MoDeL', 'ManUFacTuReR'][::-1]))

    tc.assertEqual(df_by_color.shape, df_check.shape)
    tc.assertEqual(df_by_color.columns.tolist(), df_check.columns.tolist())
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

---

### 🎯 Part 26: Sort by number of registered vehicles

#### 👇 Tasks

- ✔️ Sort `df_by_color` by `registrations` in **descending** order.
- ✔️ Update `df_by_color` without creating a new variable.
    - In other words, you're performing an **in-place** sort.

#### 🔑 Expected Output

Your index column may contain different values.

|      | manufacturer   | model   | color   |   registrations |
|-----:|:---------------|:--------|:--------|----------------:|
| 8387 | Toyota         | Camry   | Black   |           46871 |
| 8428 | Toyota         | Camry   | Silver  |           29727 |
| 8436 | Toyota         | Camry   | White   |           23874 |
| 6922 | Nissan         | Altima  | Black   |           18962 |
| 8566 | Toyota         | Corolla | Silver  |           18361 |

In [54]:
### BEGIN SOLUTION
df_by_color.sort_values('registrations', ascending=False, inplace=True)
### END SOLUTION

df_by_color.head(5)

manufacturer   model   color  registrations
7324       Toyota   Camry   Black          43233
7363       Toyota   Camry  Silver          26359
7371       Toyota   Camry   White          22224
6038       Nissan  Altima   Black          18185
7342       Toyota   Camry    Gray          16563

#### 🧭 Check Your Work

- Once you're done, run the code cell below to test correctness.
- ✔️ If the code cell runs without an error, you're good to move on.
- ❌ If the code cell throws an error, go back and fix incorrect parts.

In [55]:
# Code Generated by LambdaGrader
try:
    _did_pass = True
    _message = ''
    # DO NOT CHANGE THE CODE IN THIS CELL
    _test_case = "part-26"
    _points = 3

    df_check = df_v_backup.groupby(['COLOR', 'MODEL', 'MAKE'][::-1]) \
        .agg({'number_of_trips'.upper(): ''.join(['c', 'ou', 'n', 't'])}) \
        .reset_index()
    df_check.columns = list(map(lambda x: x.lower(), ['RegIstraTioNs', 'CoLoR', 'MoDeL', 'ManUFacTuReR'][::-1]))

    pd.testing.assert_series_equal(df_check['registrations'].sort_values(ascending=False).reset_index(drop=True),
                                   df_by_color['registrations'].reset_index(drop=True))
    pd.testing.assert_frame_equal(df_check
                                      .sort_values(['manufacturer', 'model', 'color'])
                                      .reset_index(drop=True),
                                  df_by_color
                                      .sort_values(['manufacturer', 'model', 'color'])
                                      .reset_index(drop=True))
except BaseException as ex:
    _did_pass = False
    _message = ''.join(traceback.TracebackException.from_exception(ex).format())
finally:
    _record_test_case(_test_case, _did_pass, _points, _message)

In [56]:
# LambdaGrader After File Code
# REMOVE_IN_HTML_OUTPUT
import os

grader_output_file_name = 'lambdagrader-result.json'
grading_end_time = datetime.now()

_graded_result['grading_finished_at'] = grading_end_time.isoformat()
_graded_result['grading_duration_in_seconds'] = int((grading_end_time - grading_start_time).total_seconds())
_graded_result['learner_score'] = 0
_graded_result['total_available'] = 0
_graded_result['num_test_cases'] = len(_graded_result['results'])
_graded_result['num_passed_cases'] = 0
_graded_result['num_failed_cases'] = 0

for test_case_result in _graded_result['results']:
    _graded_result['learner_score'] += test_case_result['points']
    _graded_result['total_available'] += test_case_result['available_points']
    
    if test_case_result['pass']:
        _graded_result['num_passed_cases'] += 1
    else:
        _graded_result['num_failed_cases'] += 1

print(_graded_result)
    
with open(grader_output_file_name, 'w') as fp:
    json.dump(_graded_result, fp)

{'grading_finished_at': '2022-09-21T23:53:11.369202', 'grading_duration_in_seconds': 8, 'learner_score': 80, 'total_available': 80, 'num_test_cases': 26, 'num_passed_cases': 26, 'num_failed_cases': 0, 'results': [{'test_case_name': 'part-01', 'available_points': 2, 'points': 2, 'pass': True, 'message': ''}, {'test_case_name': 'part-02', 'available_points': 2, 'points': 2, 'pass': True, 'message': ''}, {'test_case_name': 'part-03', 'available_points': 2, 'points': 2, 'pass': True, 'message': ''}, {'test_case_name': 'part-04', 'available_points': 3, 'points': 3, 'pass': True, 'message': ''}, {'test_case_name': 'part-05', 'available_points': 3, 'points': 3, 'pass': True, 'message': ''}, {'test_case_name': 'part-06', 'available_points': 2, 'points': 2, 'pass': True, 'message': ''}, {'test_case_name': 'part-07', 'available_points': 3, 'points': 3, 'pass': True, 'message': ''}, {'test_case_name': 'part-08', 'available_points': 4, 'points': 4, 'pass': True, 'message': ''}, {'test_case_name': 